Add AudioEncoderCocoa class, an AudioToolbox implementation of WebCodec's AudioEncoder

jyavenard opened the pull request:
#### c1b41e05223901c8b5a7eba74f5c72de09c12f50
<pre>
Add AudioEncoderCocoa class, an AudioToolbox implementation of WebCodec&apos;s AudioEncoder
<a href="https://bugs.webkit.org/show_bug.cgi?id=284019">https://bugs.webkit.org/show_bug.cgi?id=284019</a>
<a href="https://rdar.apple.com/140889671">rdar://140889671</a>

Reviewed by Youenn Fablet.

We add AudioEncoderCocoa, AudioToolbox implementation of WebCodec&apos;s AudioEncoder. We only support
encoding to Opus and AAC as the framework doesn&apos;t provide encoder for Flac, mp3 and vorbis (decoder only)

Enabled WPT tests.

* LayoutTests/TestExpectations:
* LayoutTests/imported/w3c/web-platform-tests/webcodecs/audio-encoder-codec-specific.https.any-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/webcodecs/audio-encoder-config.https.any-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/webcodecs/audio-encoder-config.https.any.worker-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/webcodecs/audio-encoder.https.any-expected.txt:
* LayoutTests/platform/glib/TestExpectations: Remove passing expectations, add the remaining ones that do fail.
* LayoutTests/platform/glib/imported/w3c/web-platform-tests/webcodecs/audio-encoder-config.https.any-expected.txt: Removed.
* LayoutTests/platform/glib/imported/w3c/web-platform-tests/webcodecs/audio-encoder-config.https.any.worker-expected.txt: Removed.
* LayoutTests/platform/ios/imported/w3c/web-platform-tests/webcodecs/audio-encoder-codec-specific.https.any-expected.txt: Added. The
only failure remaining is related to different default for Opus encoding which yield slightly different values to what expected.
The test performs very rough float comparisons.
* LayoutTests/platform/ios/imported/w3c/web-platform-tests/webcodecs/audio-encoder.https.any-expected.txt: Added.
* LayoutTests/platform/mac/imported/w3c/web-platform-tests/webcodecs/audio-encoder-codec-specific.https.any-expected.txt: Added.
* LayoutTests/platform/mac/imported/w3c/web-platform-tests/webcodecs/audio-encoder.https.any-expected.txt: Added.
* Source/WebCore/Modules/webcodecs/OpusEncoderConfig.h:
* Source/WebCore/Modules/webcodecs/WebCodecsAudioEncoder.cpp:
(WebCore::isSupportedEncoderCodec): Add additional checks to ensure that the values provided in the Opus config are sane.
While this isn&apos;t probably the best place to do so, as this is common for both GStreamer and Cocoa AudioEncoder, it&apos;s the simplest.
(WebCore::isValidEncoderConfig): Make sure sampleRate and numberOfChannels aren&apos;t 0.
(WebCore::createAudioEncoderConfig): Increase code readability by using struct&apos;s named member initialisation.
(WebCore::WebCodecsAudioEncoder::configure):
(WebCore::WebCodecsAudioEncoder::encode): Both Firefox and Chromes rejects AudioData that has a different sampleRate or numberOfChannels.
While we did the same, we didn&apos;t process the errors as per spec which requires that we also change the state to Closed just before
queueing the error.

(WebCore::WebCodecsAudioEncoder::isConfigSupported):
* Source/WebCore/PlatformMac.cmake:
* Source/WebCore/SourcesCocoa.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/platform/AudioEncoder.cpp:
(WebCore::AudioEncoder::create):  Plumb AudioEncoderCocoa.
* Source/WebCore/platform/AudioEncoder.h:
* Source/WebCore/platform/AudioEncoderActiveConfiguration.h: Add default initialiser to comply with more recent clang version.
It allows to make the struct member initialisation optional.
* Source/WebCore/platform/audio/cocoa/AudioDecoderCocoa.cpp:
(WebCore::InternalAudioDecoderCocoa::initialize): Change default AudioData creation to use interleaved frames instead of planar. A few
tests has this expectations, even though this isn&apos;t mandated by the specs.
* Source/WebCore/platform/audio/cocoa/AudioEncoderCocoa.cpp: Added.
(WebCore::InternalAudioEncoderCocoa::create):
(WebCore::InternalAudioEncoderCocoa::reset):
(WebCore::InternalAudioEncoderCocoa::converter):
(WebCore::InternalAudioEncoderCocoa::queueSingleton):
(WebCore::AudioEncoderCocoa::create):
(WebCore::AudioEncoderCocoa::AudioEncoderCocoa):
(WebCore::AudioEncoderCocoa::~AudioEncoderCocoa):
(WebCore::AudioEncoderCocoa::encode):
(WebCore::AudioEncoderCocoa::flush):
(WebCore::AudioEncoderCocoa::reset):
(WebCore::AudioEncoderCocoa::close):
(WebCore::InternalAudioEncoderCocoa::InternalAudioEncoderCocoa):
(WebCore::InternalAudioEncoderCocoa::initialize):
(WebCore::InternalAudioEncoderCocoa::compressedAudioOutputBufferCallback):
(WebCore::InternalAudioEncoderCocoa::generateDecoderDescriptionFromSample const):
(WebCore::InternalAudioEncoderCocoa::activeConfiguration const):
(WebCore::InternalAudioEncoderCocoa::processEncodedOutputs):
(WebCore::InternalAudioEncoderCocoa::encode):
(WebCore::InternalAudioEncoderCocoa::flush):
(WebCore::InternalAudioEncoderCocoa::close):
* Source/WebCore/platform/audio/cocoa/AudioEncoderCocoa.h: Copied from Source/WebCore/platform/AudioEncoderActiveConfiguration.h.

Canonical link: <a href="https://commits.webkit.org/287787@main">https://commits.webkit.org/287787@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/85d240d387892bbc973a5cd8dbdead8da6b25d99

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/80832 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/354 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/34768 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/85359 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/31816 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/82943 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/371 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/8153 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/63117 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/20901 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/83901 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/176 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/73565 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/43421 "Passed tests") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/133/builds/98 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/27725 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/30273 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/71659 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/28259 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/86791 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/8059 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/5683 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/71412 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/8236 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/69401 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/70653 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/17599 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/14682 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/13617 "Passed tests") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/8021 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/13542 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/7860 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/11379 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/9666 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

youennf commented on Source/WebCore/Modules/webcodecs/OpusEncoderConfig.h:
```diff
@@ -54,7 +54,7 @@ struct OpusEncoderConfig {
     OpusBitstreamFormat format { OpusBitstreamFormat::Opus };
     uint64_t frameDuration { 20000 };
-    std::optional<size_t> complexity;
+    std::optional<size_t> complexity { 9 };
```
Why optional then?

jyavenard replied:
Indeed it doesn't have to be if we set a default. In theory it should be 9 of desktop and 5 is mobile. I thought it was a fingerprint issue, so set it to 9 all the time

youennf commented on Source/WebCore/Modules/webcodecs/WebCodecsAudioEncoder.cpp:
```diff
@@ -92,6 +102,12 @@ static bool isValidEncoderConfig(const WebCodecsAudioEncoderConfig& config)
+    if (!config.sampleRate || !config.numberOfChannels)
+        return false;
+
+    if (config.bitrate && *config.bitrate > std::numeric_limits<int>::max())
```
Do we need this fix or should we instead update the WebIDL to add EnforceRange on the various properties instead?

jyavenard replied:
That's what the spec states indeed. However this is existing, I could add this as a follow-up

jyavenard replied:
the IDL already has them as [EnforceRange] but this is checking that it's not > int::max() ; while the spec has it has a "long long"  I will lodge a WPT bug.

jyavenard replied:
https://github.com/web-platform-tests/wpt/issues/49634

youennf commented on Source/WebCore/Modules/webcodecs/WebCodecsAudioEncoder.cpp:
```diff
@@ -84,6 +85,15 @@ static bool isSupportedEncoderCodec(const StringView& codec)
         return false;
 
+    // Can't find restrictions for those values, but this is what Firefox/Gecko does and WPT audio-encoder-config.https.any.html checks for.
+    if (config.sampleRate < 3000 || config.sampleRate > 384000)
```
I would be tempted to file a bug and do that as a follow-up.

jyavenard replied:
then we will fail both wpt tests.

jyavenard replied:
https://github.com/web-platform-tests/wpt/issues/49635

youennf commented on Source/WebCore/platform/audio/cocoa/AudioDecoderCocoa.cpp:
```diff
@@ -288,7 +289,7 @@ String InternalAudioDecoderCocoa::initialize(const String& codecName, const Audi
     m_inputFormatDescription = createAudioFormatDescription(*m_inputDescription, m_codecDescription.span());
 
-    m_outputDescription = CAAudioStreamDescription { double(config.sampleRate), uint32_t(config.numberOfChannels), AudioStreamDescription::Float32, CAAudioStreamDescription::IsInterleaved::No };
+    m_outputDescription = CAAudioStreamDescription { double(config.sampleRate), uint32_t(config.numberOfChannels), AudioStreamDescription::Float32, CAAudioStreamDescription::IsInterleaved::Yes };
```
We should probably file a spec bug, can you do that?

jyavenard replied:
the spec lets the UA choose, there's no option to tell what you want out which is unfortunate.
In theory, how values are packed internally should be an implementation details and it should behave the same regardless.

However, `copyTo` has a very different behaviour if the origin is interleaved or not where a planeIndex of 0 now means "all planes"

jyavenard replied:
I've opened https://github.com/w3c/webcodecs/issues/859

youennf commented on Source/WebCore/platform/audio/cocoa/AudioEncoderCocoa.cpp:
```diff
@@ -0,0 +1,385 @@
+            asbd.mFramesPerPacket = m_config.sampleRate * m_frameDuration->toDouble();
+        m_outputDescription = asbd;
+    }
+    m_codecString = codecName.isolatedCopy();
```
Do we really need this, or could we use an enumeration instead?
That could prevent issues with Strings...

jyavenard replied:
I'm happy to change this as a follow-up, right now that's pre-existing and how the GStreamer implementation uses it. Don't want to touch it here.

jyavenard replied:
we probably don't need the isolatedCopy but we generate the `activeConfiguration` on the encode workqueue. So the strings needs to work across threads.

youennf reviewed (commented):
Some comments below, I was not able to finish the review.

youennf commented on Source/WebCore/platform/audio/cocoa/AudioEncoderCocoa.cpp:
```diff
@@ -0,0 +1,386 @@
+
+    bool m_isClosed WTF_GUARDED_BY_CAPABILITY(queueSingleton()) { false };
+    const AudioEncoder::Config m_config;
+    String m_codecString; // set during initialisation, const after.
```
It is not great to use strings with work queues, I would tend to use an enum instead.

jyavenard replied:
I had planned to do it in a followup

jyavenard replied:
I've removed it fully, WebCodecsAudioEncoder turned out to copy the original string if none is provided in the new description

youennf commented on Source/WebCore/platform/audio/cocoa/AudioEncoderCocoa.cpp:
```diff
@@ -0,0 +1,386 @@
+    std::optional<MediaTime> m_frameDuration; // set during initialisation, const after.
+    std::optional<unsigned> m_complexity; // set during initialisation, const after.
+    std::optional<unsigned> m_packetlossperc; // set during initialisation, const after.
+    std::optional<bool> m_useinbandfec; // set during initialisation, const after.
```
It would be better to have a function that returns all these parameters and then we create the InternalAudioEncoderCocoa.
That way, we can make them const.

jyavenard replied:
that was more effort than worth imho.
but done

youennf commented on Source/WebCore/platform/audio/cocoa/AudioEncoderCocoa.cpp:
```diff
@@ -0,0 +1,386 @@
+    std::optional<unsigned> m_packetlossperc; // set during initialisation, const after.
+    std::optional<bool> m_useinbandfec; // set during initialisation, const after.
+    std::optional<CAAudioStreamDescription> m_inputDescription WTF_GUARDED_BY_CAPABILITY(queueSingleton());
+    bool m_isADTS { false };
```
Probably set at init time. Maybe add the same comment or make it atomic otherwise.

jyavenard replied:
not used yet. I'll add a link to the bwo

jyavenard replied:
I've removed it completely for now

youennf commented on Source/WebCore/platform/audio/cocoa/AudioEncoderCocoa.cpp:
```diff
@@ -0,0 +1,386 @@
+        CMBlockBufferRef rawBuffer = PAL::CMSampleBufferGetDataBuffer(cmSample.get());
+        ASSERT(rawBuffer);
+        RetainPtr buffer = rawBuffer;
+        // Make sure block buffer is contiguous.
```
We have this pattern in a few place, it might be handy to add a utility someday.

youennf commented on Source/WebCore/platform/audio/cocoa/AudioEncoderCocoa.cpp:
```diff
@@ -0,0 +1,386 @@
+        if (!PAL::CMBlockBufferIsRangeContiguous(rawBuffer, 0, 0)) {
+            CMBlockBufferRef contiguousBuffer;
+            if (auto error = PAL::CMBlockBufferCreateContiguous(nullptr, rawBuffer, nullptr, nullptr, 0, 0, 0, &contiguousBuffer)) {
+                RELEASE_LOG_ERROR(MediaStream, "Couldn't create buffer with error %d", error);
```
Shouldn't we error the encoder in this case (and below) by rejecting the encode native promise?

jyavenard replied:
I don't think it can ever happen. It will OOM first

jyavenard replied:
done

youennf commented on Source/WebCore/platform/audio/cocoa/AudioEncoderCocoa.cpp:
```diff
@@ -0,0 +1,386 @@
+        return EncodePromise::createAndReject("Couldn't retrieve AudioData's basic description"_s);
+
+    if (*asbd != m_inputDescription) {
+        if (m_converter)
```
```suggestion
        if (auto converter = std::exchange(m_converter, { }))
```

youennf commented on Source/WebCore/platform/audio/cocoa/AudioEncoderCocoa.cpp:
```diff
@@ -0,0 +1,386 @@
+
+    InternalAudioEncoderCocoa::queueSingleton().dispatch([weakEncoder = ThreadSafeWeakPtr { *encoder }] {
+        if (auto strongEncoder = weakEncoder.get())
+            strongEncoder->processEncodedOutputs();
```
It is not clear to me that `processEncodedOutputs()` will be executed before we resolve the encode native promise. Maybe we should keep the native promise producer somewhere and resolve/reject as part of processEncodedOutputs (and handle the case where processEncodedOutputs is not called).

jyavenard replied:
This is always called before because we return the promise itself which is resolved after the frames have returned. Same pattern as AudioDecoder

jyavenard replied:
I added a call to processEncodedOutput() at the completion of the AudioSampleBufferConverter promise to remove all doubts.

youennf commented on Source/WebCore/platform/audio/cocoa/AudioEncoderCocoa.cpp:
```diff
@@ -0,0 +1,386 @@
+        return GenericPromise::createAndResolve(); // No frame encoded yet, nothing to flush;
+    return converter()->drain()->whenSettled(queueSingleton(), [protectedThis = Ref { *this }] {
+        assertIsCurrent(queueSingleton());
+        protectedThis->processEncodedOutputs();
```
If this fails, maybe we should fail the flush promise.

jyavenard replied:
This promise is a no op in practice it can’t fail

jyavenard replied:
done

jyavenard committed:
Add AudioEncoderCocoa class, an AudioToolbox implementation of WebCodec's AudioEncoder
https://bugs.webkit.org/show_bug.cgi?id=284019
rdar://140889671

Reviewed by Youenn Fablet.

We add AudioEncoderCocoa, AudioToolbox implementation of WebCodec's AudioEncoder. We only support
encoding to Opus and AAC as the framework doesn't provide encoder for Flac, mp3 and vorbis (decoder only)

Enabled WPT tests.

* LayoutTests/TestExpectations:
* LayoutTests/imported/w3c/web-platform-tests/webcodecs/audio-encoder-codec-specific.https.any-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/webcodecs/audio-encoder-config.https.any-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/webcodecs/audio-encoder-config.https.any.worker-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/webcodecs/audio-encoder.https.any-expected.txt:
* LayoutTests/platform/glib/TestExpectations: Remove passing expectations, add the remaining ones that do fail.
* LayoutTests/platform/glib/imported/w3c/web-platform-tests/webcodecs/audio-encoder-config.https.any-expected.txt: Removed.
* LayoutTests/platform/glib/imported/w3c/web-platform-tests/webcodecs/audio-encoder-config.https.any.worker-expected.txt: Removed.
* LayoutTests/platform/ios/imported/w3c/web-platform-tests/webcodecs/audio-encoder-codec-specific.https.any-expected.txt: Added. The
only failure remaining is related to different default for Opus encoding which yield slightly different values to what expected.
The test performs very rough float comparisons.
* LayoutTests/platform/ios/imported/w3c/web-platform-tests/webcodecs/audio-encoder.https.any-expected.txt: Added.
* LayoutTests/platform/mac/imported/w3c/web-platform-tests/webcodecs/audio-encoder-codec-specific.https.any-expected.txt: Added.
* LayoutTests/platform/mac/imported/w3c/web-platform-tests/webcodecs/audio-encoder.https.any-expected.txt: Added.
* Source/WebCore/Modules/webcodecs/OpusEncoderConfig.h:
* Source/WebCore/Modules/webcodecs/WebCodecsAudioEncoder.cpp:
(WebCore::isSupportedEncoderCodec): Add additional checks to ensure that the values provided in the Opus config are sane.
While this isn't probably the best place to do so, as this is common for both GStreamer and Cocoa AudioEncoder, it's the simplest.
(WebCore::isValidEncoderConfig): Make sure sampleRate and numberOfChannels aren't 0.
(WebCore::createAudioEncoderConfig): Increase code readability by using struct's named member initialisation.
(WebCore::WebCodecsAudioEncoder::configure):
(WebCore::WebCodecsAudioEncoder::encode): Both Firefox and Chromes rejects AudioData that has a different sampleRate or numberOfChannels.
While we did the same, we didn't process the errors as per spec which requires that we also change the state to Closed just before
queueing the error.

(WebCore::WebCodecsAudioEncoder::isConfigSupported):
* Source/WebCore/PlatformMac.cmake:
* Source/WebCore/SourcesCocoa.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/platform/AudioEncoder.cpp:
(WebCore::AudioEncoder::create):  Plumb AudioEncoderCocoa.
* Source/WebCore/platform/AudioEncoder.h:
* Source/WebCore/platform/AudioEncoderActiveConfiguration.h: Add default initialiser to comply with more recent clang version.
It allows to make the struct member initialisation optional.
* Source/WebCore/platform/audio/cocoa/AudioDecoderCocoa.cpp:
(WebCore::InternalAudioDecoderCocoa::initialize): Change default AudioData creation to use interleaved frames instead of planar. A few
tests has this expectations, even though this isn't mandated by the specs.
* Source/WebCore/platform/audio/cocoa/AudioEncoderCocoa.cpp: Added.
(WebCore::InternalAudioEncoderCocoa::create):
(WebCore::InternalAudioEncoderCocoa::reset):
(WebCore::InternalAudioEncoderCocoa::converter):
(WebCore::InternalAudioEncoderCocoa::queueSingleton):
(WebCore::AudioEncoderCocoa::create):
(WebCore::AudioEncoderCocoa::AudioEncoderCocoa):
(WebCore::AudioEncoderCocoa::~AudioEncoderCocoa):
(WebCore::AudioEncoderCocoa::encode):
(WebCore::AudioEncoderCocoa::flush):
(WebCore::AudioEncoderCocoa::reset):
(WebCore::AudioEncoderCocoa::close):
(WebCore::InternalAudioEncoderCocoa::InternalAudioEncoderCocoa):
(WebCore::InternalAudioEncoderCocoa::initialize):
(WebCore::InternalAudioEncoderCocoa::compressedAudioOutputBufferCallback):
(WebCore::InternalAudioEncoderCocoa::generateDecoderDescriptionFromSample const):
(WebCore::InternalAudioEncoderCocoa::activeConfiguration const):
(WebCore::InternalAudioEncoderCocoa::processEncodedOutputs):
(WebCore::InternalAudioEncoderCocoa::encode):
(WebCore::InternalAudioEncoderCocoa::flush):
(WebCore::InternalAudioEncoderCocoa::close):
* Source/WebCore/platform/audio/cocoa/AudioEncoderCocoa.h: Copied from Source/WebCore/platform/AudioEncoderActiveConfiguration.h.

Canonical link: https://commits.webkit.org/287787@main
